p4-fusion/utils/arguments.cc

```
@@ -7,17 +7,93 @@
+static std::unique_ptr<std::vector<std::string>> ArgsFromFile(const std::string& filename)
+{
+	std::ifstream infile(filename);
+	if (!infile.is_open())
+	{
+		ERR("Failed to open file: " << filename);
+		return nullptr;
+	}
+	std::string line;
+	std::stringstream ss;
+	while (std::getline(infile, line))
+	{
+		// Ignore empty lines and comments
+		if (!line.empty() && line[0] != '#')
+		{
+			ss << line << " ";
+		}
+	}
+	std::string fullcontent = ss.str();
+	wordexp_t expRes;
+	int resultCode = wordexp(fullcontent.c_str(), &expRes, WRDE_NOCMD);
+	if (resultCode)
+	{
+		std::string errorMessage;
+		switch (resultCode)
+		{
+			case WRDE_BADCHAR: errorMessage = "Illegal newline or |&;<>(){}"; break; 
+			case WRDE_BADVAL: errorMessage = "An undefined shell variable was referenced"; break;
+			case WRDE_CMDSUB: errorMessage = "Command substitution not allowed"; break;
+			case WRDE_NOSPACE: errorMessage = "Out of memory"; break;
+			case WRDE_SYNTAX: errorMessage = "Syntax error"; break;
+			default: break;
+		}
+		ERR("Error parsing config file: " << errorMessage);
+		return nullptr;
+	}
+
+	std::unique_ptr<std::vector<std::string>> args(new std::vector<std::string>());
+	for (size_t i = 0; i < expRes.we_wordc; i++)
+	{
+		args->emplace_back(expRes.we_wordv[i]);
+	}
+	wordfree(&expRes);
+	return args;
+}
 
 void Arguments::Initialize(int argc, char** argv)
 {
+	m_Parameters.emplace("--config", ParameterData { false, false, {}, "Path to a file that can contain additional command line arguments." });
 	for (int i = 1; i < argc - 1; i += 2)
 	{
-		std::string name = argv[i];
+		const std::string name = argv[i];
+		const std::string value = argv[i + 1];
 
-		if (m_Parameters.find(name) != m_Parameters.end())
+		const auto it = m_Parameters.find(name);
+		if (it != m_Parameters.end())
 		{
-			m_Parameters.at(name).valueList.push_back(argv[i + 1]);
-			m_Parameters.at(name).isSet = true;
+			ParameterData& param = it->second;
+			if (name == "--config" && !param.isSet)
```

Could you please explain the purpose of `!param.isSet`?

---

This way you can only specify one config file. There are valid reasons for being able to specify multiple, though, if that's something we need.

---

I think we should either handle this as an error, or allow multiple files. While for "regular" command line parameters it does not come as a surprise that we expect the user to specify exactly once, but in the case of response files, it's not as evident.

IMHO the worst we can do is "silently" ignore further config files.

---

The existing logic for regular arguments is not much better: it silently ignores all but the last value if specified more than once. Do you think that should be made into an error as well?

---

For regular arguments, the _behavior_ itself I think is correct. For response file, I'm not sure.

---

Enabled multiple config files